iwyu_include_picker.cc

```diff
@@ -1329,13 +1329,32 @@ vector<string> IncludePicker::GetCandidateHeadersForFilepathIncludedFrom(
   // ConvertToQuotedInclude because it avoids trouble when the same
   // file is accessible via different include search-paths, or is
   // accessed via a symlink.
-  const string& quoted_include_as_typed
-      = MaybeGetIncludeNameAsWritten(including_filepath, included_filepath);
-  if (!quoted_include_as_typed.empty()) {
-    vector<string>::iterator it = std::find(retval.begin(), retval.end(),
-                                            quoted_includee);
-    if (it != retval.end())
-      *it = quoted_include_as_typed;
+  vector<string>::iterator it = std::find(retval.begin(), retval.end(),
+                                          quoted_includee);
+  if (it != retval.end()) {
+      const string& quoted_include_as_typed
+          = MaybeGetIncludeNameAsWritten(including_filepath, included_filepath);
+      if (!quoted_include_as_typed.empty()) {
+          *it = quoted_include_as_typed;
+      } else {
+          string strippedIncludee = included_filepath;
+          // Directory of the file is implicitly in the search path.
+          // So if we have the full path, shorten it
+          const string implicitHeaderSearchPath =
+                  CanonicalizeHeaderSearchPath(GetParentPath(including_filepath));
+          if (StripLeft(&strippedIncludee, implicitHeaderSearchPath)) {
+              VERRS(6) << "Adding include in same directory for "
+                       << included_filepath << ": " << strippedIncludee
+                       << "\n";
+              // If the included path is absolute, we prefer the short one
+              // else we prefer the current one, which is probably relative to
+              // a search directory
+              if(IsAbsolutePath(StripQuotes(*it)))
+                  retval.insert(it, "\"" + strippedIncludee + "\"");
+              else
+                  retval.push_back("\"" + strippedIncludee + "\"");
+          }
+      }
```

Comment:
`MaybeGetIncludeNameAsWritten` is meant to be return something, if possible (hence the "maybe") I assume it should return the include line unmodified if it was already used. So IWYU does not change an include line the user has explicitly written and also to "adapt" to user style. If we did not find a user supplied include line (->New include line) we are in the else branch. So empty is on purpose.

This is NOT a fix for the leading slash, although I'm curious where it originated. The else is really for handling new headers. BTW: It is not introduced by this PR. So refactoring of this (like making the function return a bool and taking a pointer param) should be done outside of this PR.